File: identity-vault/www/ionicnativeauth.js
var exec = require('cordova/exec');

var PLUGIN_NAME = 'IonicNativeAuth';

var IonicNativeAuth = {
  getVault: function(options) {
    var instance = new AuthVault(options);
    var resolver;
    var rejecter;
    var p = new Promise(function(resolve, reject) { resolver = resolve; rejecter = reject});
    instance._setupPromise = p;
    exec(function(event) {
      instance.handle(event);
      options.onReady && options.onReady(instance);
      resolver();
    }, function(err) {
      options.onError && options.onError(err);
      rejecter();
    }, PLUGIN_NAME, 'setup', [options]);
    return instance;
  },
};


function unwrapData(key, resolve) {
  return function(data) {
    resolve(data && data[key]);
  }
}

function AuthVault(options) {
  return {
    onLock: options.onLock,
    onUnlock: options.onUnlock,
    onConfig: options.onConfig,
    descriptor: {
      username: options.username,
      vaultId: options.vaultId,
    },
    _setupPromise: null,
    config: null,
    handlerId: null,
    getConfig: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'getConfig', [descriptor]);
        });
      });
    },
    unsubscribe: function() {
      var descriptor = this.descriptor
      var handlerId = this.handlerId
      return this._setupPromise.then(function() {
        return new Promise(function (resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'close', [descriptor, handlerId]);
        });
      });
    },
    clear: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'clear', [descriptor]);
        });
      });
    },
    isLocked: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'isLocked', [descriptor]);
        });
      });
    },
    isLockedOutOfBiometrics: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'isLockedOutOfBiometrics', [descriptor]);
        });
      });
    },
    isInUse: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'isInUse', [descriptor]);
        });
      });
    },
    remainingAttempts: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'remainingAttempts', [descriptor]);
        });
      });
    },
    getToken: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(unwrapData('token', resolve), reject, PLUGIN_NAME, 'getValue', [descriptor, 'token']);
        });
      });
    },
    storeToken: function(token) {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'storeValue', [descriptor, "token", token]);
        });
      });
    },
    getValue: function(key) {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(unwrapData(key, resolve), reject, PLUGIN_NAME, 'getValue', [descriptor, key]);
        });
      });
    },
    getKeys: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'getKeys', [descriptor]);
        });
      });
    },
    storeValue: function(key, value) {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'storeValue', [descriptor, key, value]);
        });
      });
    },
    removeValue: function(key) {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'removeValue', [descriptor, key]);
        });
      });
    },
    getUsername: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'getUsername', [descriptor]);
        });
      });
    },
    lock: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'lock', [descriptor]);
        });
      });
    },
    getBiometricType: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'getBiometricType', [descriptor]);
        });
      });
    },
    getAvailableHardware: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'getAvailableHardware', [descriptor]);
        });
      });
    },
    isBiometricsAvailable: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'isBiometricsAvailable', [descriptor]);
        });
      });
    },
    isBiometricsSupported: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'isBiometricsSupported', [descriptor]);
        });
      });
    },
    setBiometricsEnabled: function(isBiometricsEnabled) {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'setBiometricsEnabled', [descriptor, isBiometricsEnabled]);
        });
      });
    },
    isBiometricsEnabled: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'isBiometricsEnabled', [descriptor]);
        });
      });
    },
    isSecureStorageModeEnabled: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'isSecureStorageModeEnabled', [descriptor]);
        });
      });
    },
    setSecureStorageModeEnabled: function(enable) {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'setSecureStorageModeEnabled', [descriptor, enable]);
        });
      });
    },
    isPasscodeEnabled: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'isPasscodeEnabled', [descriptor]);
        });
      });
    },
    setPasscodeEnabled: function(enable) {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'setPasscodeEnabled', [descriptor, enable]);
        });
      });
    },
    setPasscode: function(passcode) {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'setPasscode', [descriptor, passcode]);
        });
      });
    },
    isPasscodeSetupNeeded: function() {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'isPasscodeSetupNeeded', [descriptor]);
        });
      });
    },
    unlock: function(withPasscode, passcode) {
      var descriptor = this.descriptor
      return this._setupPromise.then(function() {
        return new Promise(function(resolve, reject) {
          exec(resolve, reject, PLUGIN_NAME, 'unlock', [descriptor, withPasscode, passcode]);
        });
      });
    },
    handle: function(event) {
      if (event && event.event) {
        switch (event.event) {
          case 'lock': {
            this.onLock && this.onLock(event.data);
            this.handlerId = event.handlerId
            break;
          }
          case 'config': {
            this.config = event.data;
            this.onConfig && this.onConfig(event.data);
            this.handlerId = event.handlerId
            break;
          }
          case 'unlock': {
            this.onUnlock && this.onUnlock(event.data);
            this.handlerId = event.handlerId
            break;
          }
        }
      }
    },
  };
}

module.exports = IonicNativeAuth;

File: identity-vault/src/ios/IonicNativeAuth.swift
@objc(IonicNativeAuthPlugin)
class IonicNativeAuthPlugin: CDVPlugin {

    private var vaultStore: [IdentityVault]?

    public override func pluginInitialize() {
        vaultStore = []
    }

    private func ensureVault(_ command: CDVInvokedUrlCommand) -> IdentityVault? {
        do {
            let descriptor = try getDescriptorFromCommand(command)
            guard let vault = getVaultFromStore(descriptor.key) else {
                sendResult(command, VaultError.vaultUnavailable)
                return nil
            }
            return vault
        } catch {
            sendResult(command, error)
            return nil
        }
    }

    private func getVaultFromStore(_ key: String) -> IdentityVault? {
        for vault in vaultStore! {
            if vault.descriptor.key == key {
                return vault
            }
        }
        return nil
    }

    private func storeVault(_ vault: IdentityVault) {
        vaultStore!.append(vault)
    }

    private func getDescriptorFromCommand(_ command: CDVInvokedUrlCommand) throws -> VaultDescriptor {
        let descriptors = command.arguments[0] as? [String:Any] ?? [:]
        guard let username = descriptors["username"] as? String else { throw VaultError.invalidArguments("missing username")}
        guard let vaultId = descriptors["vaultId"] as? String else { throw VaultError.invalidArguments("missing vaultId")}
        return VaultDescriptor(username: username, vaultId: vaultId)
    }

    @objc func setup(_ command: CDVInvokedUrlCommand) {
        let setupOptions = command.arguments[0] as? [String:Any] ?? [:]
        guard let descriptor = try? getDescriptorFromCommand(command) else {
            sendResult(command, VaultError.invalidArguments("missing username or vaultId"))
            return
        }
        var authVault = getVaultFromStore(descriptor.key)
        if authVault == nil {
            authVault = IdentityVault(setupOptions, descriptor, eventHandler: sendEvents)
            storeVault(authVault!)
        }
        guard let vault = authVault else { return }

        vault.addEventHandler(command)

        vault.sendConfigEvent()
    }

    @objc func getConfig(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        // send the config event as well as the result
        sendResult(command, vault.getConfig())
    }

    @objc func isBiometricsAvailable(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.isBiometricsAvailable())
    }

    @objc func isBiometricsSupported(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.isBiometricsSupported())
    }

    @objc func isPasscodeSetupNeeded(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.isPasscodeSetupNeeded())
    }

    @objc func getBiometricType(_ command: CDVInvokedUrlCommand) {
        // deprecated, getAvailableHardware should be preferred
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.getBiometricType().deprecatedType)
    }
    
    @objc func getAvailableHardware(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        let types: [String] = [vault.getBiometricType()].compactMap { (type) -> String? in
            return (type == .none) ? nil : type.rawValue
        }
        sendResult(command, types)
    }
    
    @objc func lock(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        vault.lock()
        sendResult(command)
    }

    @objc func isLocked(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.isLocked())
    }

    @objc func isLockedOutOfBiometrics(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.isLockedOutOfBiometrics())
    }

    @objc func isInUse(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.isInUse())
    }

    @objc func remainingAttempts(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.remainingAttempts())
    }

    @objc func getValue(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }

        guard let key = command.arguments[1] as? String else {
            return sendResult(command, VaultError.invalidArguments("key missing"))
        }

        do {
            guard let value = try vault.getStoredValue(forKey: key) else {
                return sendResult(command)
            }
            return sendResult(command, [key:value])
        } catch {
            return sendResult(command, error)
        }
    }
    
    @objc func getKeys(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        do {
            guard let value = try vault.getStoredKeys() else {
                return sendResult(command)
            }
            return sendResult(command, value)
        } catch {
            return sendResult(command, error)
        }
    }

    @objc func storeValue(_ command: CDVInvokedUrlCommand) {
        guard command.arguments.count == 3 else {
            return sendResult(command, VaultError.invalidArguments("wrong number of args, missing key or data"))
        }

        guard let key = command.arguments[1] as? String else {
            return sendResult(command, VaultError.invalidArguments("missing key"))
        }

        let data = command.arguments[2]

        guard let vault = ensureVault(command) else { return }

        do {
            try vault.storeValue(key, data)
            sendResult(command)
        } catch {
            sendResult(command, error)
        }
    }

    @objc func removeValue(_ command: CDVInvokedUrlCommand) {
        guard command.arguments.count == 2 else {
            return sendResult(command, VaultError.invalidArguments("wrong number of args, missing key or data"))
        }

        guard let key = command.arguments[1] as? String else {
            return sendResult(command, VaultError.invalidArguments("missing key"))
        }

        guard let vault = ensureVault(command) else { return }

        do {
            try vault.removeValue(key)
            sendResult(command)
        } catch {
            sendResult(command, error)
        }
    }

    @objc func getUsername(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.getUsername())
    }

    @objc func getVaultId(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.getVaultId())
    }

    @objc func setBiometricsEnabled(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        guard let enabled = command.arguments[1] as? Bool else {
            return sendResult(command, VaultError.invalidArguments("enabled arg missing or not a boolean"))
        }

        do {
            try vault.setBiometricsEnabled(enabled)
        } catch {
            return sendResult(command, error)
        }
        sendResult(command)
    }

    @objc func isBiometricsEnabled(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.isBiometricsEnabled())
    }

    @objc func setSecureStorageModeEnabled(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        guard let enabled = command.arguments[1] as? Bool else {
            return sendResult(command, VaultError.invalidArguments("enabled arg missing or not a boolean"))
        }

        do {
            try vault.setSecureStorageModeEnabled(enabled)
        } catch {
            return sendResult(command, error)
        }
        sendResult(command)
    }

    @objc func isSecureStorageModeEnabled(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.isSecureStorageModeEnabled())
    }

    @objc func isPasscodeEnabled(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        sendResult(command, vault.isPasscodeEnabled())
    }

    @objc func clear(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        do {
            try vault.clear()
            sendResult(command)
        } catch {
            return sendResult(command, error)
        }
    }

    @objc func close(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        guard let handlerId = command.arguments[1] as? String else {
            return sendResult(command, VaultError.invalidArguments("handlerId missing"))
        }
        vault.removeEventHandler(handlerId)
        if vault.getHandlers().count == 0 {
        }
        sendResult(command)
    }

    @objc func setPasscodeEnabled(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        guard let enabled = command.arguments[1] as? Bool else {
            return sendResult(command, VaultError.invalidArguments("enabled arg missing or not a boolean"))
        }

        do {
            try vault.setPasscodeEnabled(enabled)
        } catch {
            return sendResult(command, error)
        }
        sendResult(command)
    }

    @objc func setPasscode(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        let passcode = command.arguments[1] as? String

        do {
            let wasNeeded = vault.isPasscodeSetupNeeded()
            try vault.setPasscode(passcode) { error in
                guard let err = error else {
                    if (wasNeeded) { vault.sendConfigEvent() }
                    return self.sendResult(command)
                }
                return self.sendResult(command, err)
            }
        } catch {
            return sendResult(command, error)
        }
    }

    @objc func unlock(_ command: CDVInvokedUrlCommand) {
        guard let vault = ensureVault(command) else { return }
        let withPasscode = command.arguments[1] as? Bool ?? false
        let passcode = command.arguments[2] as? String
        if (!vault.isLocked()) { return sendResult(command) }
        do {
            try vault.unlock(withPasscode, passcode) { error in
                guard let err = error else {
                    vault.sendEvent("unlock", vault.getConfig())
                    return self.sendResult(command)
                }
                return self.sendResult(command, err)
            }
        } catch {
            return sendResult(command, error)
        }
    }

    private func sendEvents(_ event: [String:Any] = [:], _ vault: IdentityVault) {
        for command in vault.getHandlers() {
            var eventToSend: [String:Any] = ["handlerId":command.callbackId]
            for (k, v) in event {
                eventToSend[k] = v
            }
            sendResult(command, eventToSend, keepCallback: true)
        }
    }

    private func sendResult(_ command: CDVInvokedUrlCommand, _ error: Error) {
        if let descriptor = try? getDescriptorFromCommand(command) {
            if let vault = getVaultFromStore(descriptor.key) {
                switch (error) {
                case VaultError.tooManyFailedAttempts:
                    if vault.config.shouldClearVaultAfterTooManyFailedAttempts {
                        try? vault.clear()
                    }
                    break;
                default:
                    break;
                }
            }
        }
        
        let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: getVaultErrorObject(error).asDictionary)
        result?.setKeepCallbackAs(true)
        commandDelegate.send(result, callbackId: command.callbackId)
    }

    private func sendResult(_ command: CDVInvokedUrlCommand, keepCallback: Bool = false) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK)
        result?.setKeepCallbackAs(keepCallback)
        commandDelegate.send(result, callbackId: command.callbackId)
    }

    private func sendResult(_ command: CDVInvokedUrlCommand, _ data: [String:Any], keepCallback: Bool = false) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: data)
        result?.setKeepCallbackAs(keepCallback)
        commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    private func sendResult(_ command: CDVInvokedUrlCommand, _ data: [String], keepCallback: Bool = false) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: data)
        result?.setKeepCallbackAs(keepCallback)
        commandDelegate.send(result, callbackId: command.callbackId)
    }

    private func sendResult(_ command: CDVInvokedUrlCommand, _ data: String) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: data)
        commandDelegate.send(result, callbackId: command.callbackId)
    }

    private func sendResult(_ command: CDVInvokedUrlCommand, _ data: Int) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: data)
        commandDelegate.send(result, callbackId: command.callbackId)
    }

    private func sendResult(_ command: CDVInvokedUrlCommand, _ data: Bool) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: data)
        commandDelegate.send(result, callbackId: command.callbackId)
    }
}
